### PR TITLE
rule: react/filename-extension

### PR DESCRIPTION
**Pull request by ematipico**
This PR implements the rule [jsx-filename-extension](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md).

It's still a work in progress because the current implementation doesn't check function components.
I was wondering if it would be useful to have an utility function that checks that

### CODE REVIEW

File: packages/@romejs/compiler/lint/rules/react/filenameExtension.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+function isClassComponent(node: AnyNode, path: Path) {
+	return (
+		(node.type === "JSClassHead" &&
+		node.superClass !== undefined &&
+		doesNodeMatchReactPattern(node, path.scope, "React.Component")) ||
+		doesNodeMatchReactPattern(node, path.scope, "Component") ||
+		doesNodeMatchReactPattern(node, path.scope, "React.PureComponent") ||
+		doesNodeMatchReactPattern(node, path.scope, "PureComponent")
+	);
+}
```
Comment (sebmck) [inline packages/@romejs/compiler/lint/rules/react/filenameExtension.ts:17]: We might have a util method already for this. @bitpshr or @VictorHom might know.

File: packages/@romejs/compiler/lint/rules/react/filenameExtension.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+export default {
+	name: "reactFilenameExtension",
+	enter(path: Path): TransformExitResult {
+		const {node, context} = path;
+
+		const extension = context.path.getExtensions();
+		if (
+			isClassComponent(node, path) &&
+			ALLOWED_EXTENSIONS.includes(extension) === false
```
Comment (sebmck) [inline packages/@romejs/compiler/lint/rules/react/filenameExtension.ts:27]: `context.path.getExtensions()` is cheap (it's memoized) but we should ideally do it lazily.

```suggestion
		if (
			isClassComponent(node, path) &&
			ALLOWED_EXTENSIONS.includes(context.path.getExtensions()) === false
```

**Comment by sebmck** [comment]
Looks good! Thank you!

File: packages/@romejs/compiler/lint/rules/react/filenameExtension.ts (hunk 1 above)
Comment (VictorHom) [reply packages/@romejs/compiler/lint/rules/react/filenameExtension.ts:17]: I haven't used it yet myself, (credit to @bitpshr for the utilities), but `packages/@romejs/compiler/lint/utils/react/insideClassComponent.ts` is where the utility lives

Comment (ematipico) [reply packages/@romejs/compiler/lint/rules/react/filenameExtension.ts:17]: I tried to use that method but the snapshot gave me 4 errors. The utility loops over the ancestors until it finds a match. Probably it should be reviewed or I didn't use it correctly

**Comment by ematipico** [comment]
> I feel like this rule should not check for React at all, and should instead only check for the presence of JSX (regardless of whether it's a component or not.)

You're totally right!  I will review the logic, make it agnostic and move it under 'jsx-a11y' category. It makes more sense

**Comment by sebmck** [comment]
This isn't related to accessibility so we might want just a `jsx` lint category. I can imagine some other rules we'd want to have there.

**Comment by bitpshr** [comment]
@sebmck agreed; lots of the rules in the `react` package are really JSX-specific. I’ll open a PR to reorganize accordingly. This is another deviation from the upstream ESLint rules and their organization that I think makes sense and improves clarity.

**Comment by bitpshr** [comment]
Related: https://github.com/romejs/rome/pull/675

File: packages/@romejs/compiler/lint/rules/react/filenameExtension.ts (hunk 1 above)
Comment (ematipico) [reply packages/@romejs/compiler/lint/rules/react/filenameExtension.ts:17]: Done! 

FYI the utility has to be used in pair with another check:

```js
node.type === "JSClassHead" && insideClassComponent(path)
```

**Comment by bitpshr** [comment]
@ematipico I went ahead and cleaned up a general JSX-catching rule that I had locally and pushed it up at https://github.com/romejs/rome/pull/680. The approach there isn't tied to React components and should be a bit more agnostic: it checks for any JSX element, and raises a diagnostic if the current file extension is not `.jsx` or `.tsx`. Let me know what you think.

**Comment by ematipico** [comment]
Yeah I'd say we can close this one in favor of yours!

**Pull request closed without merging**